### PR TITLE
fix: Enable Pipettes for Flex

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

- Add ability to specify pipettes for OT-3
- Refactor OT-2 pipette specification
- Make specifying pipettes in the configuration file nicer

Users will now specify pipettes as follows:

**OT-2**

```yaml
robot:
  id: otie
  hardware: ot2
  emulation-level: hardware
  hardware-specific-attributes:
    right-pipette: P20 Single
    left-pipette: P300 Multi
```

**OT-3 (96 Channel)**

```yaml
robot:
  id: otie
  hardware: ot3
  emulation-level: hardware
  hardware-specific-attributes:
    right-pipette: P1000 96 Channel
```

**OT-3 (Not 96 Channel)**

```yaml
robot:
  id: otie
  hardware: ot3
  emulation-level: hardware
  hardware-specific-attributes:
    right-pipette: P1000 Multi
    left-pipette: P50 Single
```

# Remaining Tasks

- [x] Add logic to prevent user from specifing 96-channel in RIGHT mount
- [x] Fix unit tests that broke with this change
- [x] Fix e2e tests that broke with this change
- [x] Add eeprom volumes for gripper, left-pipette, and right-pipette. These allow the ot3-firmware-builder to push eeprom.bin files to the correct containers
- [ ] Inside of monorepo add function to generate pipette serial number string from `OT3_PIPETTE_DEFINITION` and `EEPROM_FILENAME` environment variables.
  - [ ] Have python function write to a passed eeprom file
  - [ ] Delete file contents first
  - [ ] Print serial string to console for debug purposes
  - [ ] Add unit tests to cover this
  - [ ] Merge to edge and test during a `refresh-dev`

- [ ] Add e2e test to verify that eeprom files have correct content
- [ ] Build in OT-2 pipette serial generation instead of requiring user to specify
  - [ ] Add unit tests to cover this
- [ ] Update sample files with pipettes
- [ ] Update documentation with new pipette specification model and valid pipette names
- [ ] Add documentation for how to update the pipette_version.json

# Test Plan

- [ ] All OT-2 pipettes can be specified 
- [ ] All OT-3 (Flex) pipettes can be specified
- [ ] All unit tests pass
- [ ] All E2E tests pass
- [ ] Emulated OT-2 starts with specified pipettes
- [ ] Emulated OT-3 starts with specified pipettes
- [ ] Validate functionality inside .devcontainer with @y3rsh 

# Changelog

A lot of spread out changes in this one

## `pipette_utils` package

This package provides opentrons-emulation with the following: 
- The ability to define what pipettes are available for selection for both the OT-2 and OT-3 (Flex). 
- The parameters required to generate a serial number
- Environment Variable specification to share above parameters with necessary containers

It is split up into 2 files, `lookups.py` and `data_models.py`.

Defined inside `__init__.py`, the public function `get_robot_pipettes,` exposes the package's functionality. 

### lookups.py

Defines pipettes available for selection in emulation. It is split up into 2 enums, `OT2PipetteLookup` and `OT3PipetteLookup` which both inherit common methods from `BasePipetteLookup`

Each pipette is defined as an entry in the one of the enums

### data_models.py

There are 2 dataclasses inside of data_models that comprise of all the information necessary to define a robot's pipettes. 

`PipetteInfo` 

Contains all necessary information for a pipette

- Can be instantiated with 

`RobotPipettes` 

Wraps a left and right `PipetteInfo` into a single class

# Review requests

- Does the paradigm I created for pipettes make sense? 

# Risk assessment

Medium, could break OT-2 pipette specification. Will test around that though

### CODE REVIEW

File: emulation_system/emulation_system/compose_file_creator/input/hardware_models/robots/ot3_model.py [hunk 1]
```diff
@@ -24,10 +24,10 @@
 
 
 class OT3Attributes(HardwareSpecificAttributes):
-    """Attributes specific to OT3."""
+    """Attributes specific to Robots."""
 
-    left: PipetteSettings = Field(default=PipetteSettings())
-    right: PipetteSettings = Field(default=PipetteSettings())
+    left_pipette: OT3Pipettes.valid_pipette_name_enum() | None = None  # type: ignore[valid-type]
```
Comment (sfoster1) [inline emulation_system/emulation_system/compose_file_creator/input/hardware_models/robots/ot3_model.py:29]: I wish mypy allowed this but they want like "static analysis for safety and speed" or whatever

File: emulation_system/emulation_system/compose_file_creator/pipette_utils.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    # We want to construct an enum whose values are the pipette_name values from classes that subclass
+    # this class
+    # This method is called inside of input/hardware_models/robots/ot2_model.py and
+    # input/hardware_models/robots/ot3_model.py. Mypy is not happy with this because it doesn't
+    # like the fact that we are calling a method to define the types.
+    # So we are address this by telling mypy to be quiet by using # type: ignore[valid-type] and
+    # moving on with our very long day.
+
+    @classmethod
+    def _get_pipette_names(cls) -> Generator[Tuple[str, str], None, None]:
```
Comment (sfoster1) [inline emulation_system/emulation_system/compose_file_creator/pipette_utils.py:98]: nasty, i love it

**Review by sfoster1 — APPROVED** [review]
approach definitely looks good to me!

File: emulation_system/emulation_system/compose_file_creator/pipette_utils.py (hunk 2 above)
Comment (fsinapi) [inline emulation_system/emulation_system/compose_file_creator/pipette_utils.py:98]: This is very neat and the explanation is very helpful

**Review by fsinapi — APPROVED** [review]
This approach LGTM!

**Comment by DerekMaggio** [comment]
Closing in favor of smaller, more targeted PRs. 
Will still be pulling code from this branch

**Pull request closed without merging**